File: 08_김채린/4949.c
#include <stdio.h>
int i, top, n, stack[1000];
char str[1000];

void push() { 
    top++; 
    stack[top] = str[i]; 
}

int pop() { 
    //닫는 괄호가 현재 문자열(str)에 있고, 시작하는 괄호가 스택(stack)에 존재한다면
    if ((str[i] == ')' && stack[top] == '(') || (str[i] == ']' && stack[top] == '[')) {
        top--;
        return 1;
    } 
    else 
        return 0; 
}

int main() {
    while (1) {
        scanf("%s", str);
        top = -1;
        n = 1;
        if (str[0] == '.') 
            break;
        for (i = 0; str[i] != '.'; i++) {
            if (!n) 
                continue;
            if (str[i] == '(') //중괄호가 시작하면 스택에 삽입, stack[top]='('
                push();
            if (str[i] == '[') //대괄호가 시작하면 스택에 삽입, stack[top]='['
                push();
            if (str[i] == ']') //대괄호가 끝나면 스택에서 pop, n=1
                n = pop();
            if (str[i] == ')') //중괄호가 끝나면 스택에서 pop, n=1
                n = pop();
        }
        if (n && top == -1) //top이 초기화되어있고(스택이 비어있고), n이 1이라면
            puts("yes");
        else 
            puts("no");
    }
}

File: 09_민새미/4949.py
# 하나 또는 여러줄에 걸쳐서 문자열이 주어진다.
# 각 문자열은 영문 알파벳, 공백, 소괄호("( )") 대괄호("[ ]")등으로 이루어져 있으며, 길이는 100글자보다 작거나 같다.
# 입력의 종료조건으로 맨 마지막에 점 하나(".")가 들어온다.

#yes가 계속 안나오게 하는 법...어떻게 하는거지

stack =[]
while(1):
    i = input()

    if(i[-1]!='.'):
        print("입력의 종료조건으로 맨 마지막에 점이 들어와야합니다!")
        print("no")
        
    else:

        for x in i:
            #print(x)
            if x =='{' or x== '[' or x=='(' :
                # print(x)
                stack.append(x)

            elif x=="]":
                if(len(stack)<1):
                    print("no")
                else:
                    top=stack.pop()
                    if(top!="["):
                        print("no")
                        # print("괄호가 맞지 않습니다")
                        
                        break
                    else:
                        print("yes")
                        # print("괄호 잘 맞췄습니다")
                        
            elif x==")":
                if(len(stack)<1):
                    print("no")
                else:
                    top=stack.pop()
                    if(top!="("):
                        print("no")
                        # print("괄호가 맞지 않습니다")
                        break
                    else:
                        print("yes")
                        # print("괄호 잘 맞췄습니다")
        stack=[]    


# while True :
#     a = input()
#     stack = []

#     if a == "." :
#         break

#     for i in a :
#         if i == '[' or i == '(' :
#             stack.append(i)
#         elif i == ']' :
#             if len(stack) != 0 and stack[-1] == '[' :
#                 stack.pop() # 맞으면 지워서 stack을 비워줌 0 = yes
#             else : 
#                 stack.append(']')
#                 break
#         elif i == ')' :
#             if len(stack) != 0 and stack[-1] == '(' :
#                 stack.pop()
#             else :
#                 stack.append(')')
#                 break
#     if len(stack) == 0 :
#         print('yes')
#     else :
#         print('no')
        